cmd/slareport.go

```diff
@@ -0,0 +1,129 @@
+var (
+	SlareportCMD       = newSlareportCMD()
+	customer           string
+	date               string
+	timeRange          string
+	promURL            string
+	awssecretaccesskey string
+	awsaccesskeyid     string
+	endpointURL        string
+	bucket             string
+	mimirOrg           string
+	previousMonth      bool
+)
+
+func newSlareportCMD() *cobra.Command {
+
+	command := &cobra.Command{
+		Use:   "slareport",
+		Short: "Generate SLA reports for customers",
+		Long: `
+Run the sla report for services. This will query a prometheus endpoint, generate the SLA reports and
+upload them to an S3 bucket.`,
+		RunE: c.runReport,
+	}
+
+	now := time.Now()
+
+	viper.AutomaticEnv()
+	viper.SetDefault("PROM_URL", "http://localhost:9090/prometheus")
+	viper.SetDefault("BUCKET_NAME", "slareports")
+	viper.SetDefault("AWS_SECRET_ACCESS_KEY", "minioadmin")
+	viper.SetDefault("AWS_ACCESS_KEY_ID", "minioadmin")
+	viper.SetDefault("ENDPOINT_URL", "http://localhost:9000")
+
+	command.Flags().StringVar(&customer, "customer", "", "Set the customer name to run against a specific customer")
+	command.Flags().StringVar(&date, "date", now.Format(time.RFC3339), "Specify the date from which the SLA should be reported")
+	command.Flags().StringVar(&timeRange, "range", "30d", "Range during which the sla should be evaluated")
+	command.Flags().StringVar(&promURL, "promurl", viper.GetString("PROM_URL"), "Prometheus URL where the SLA metrics reside. ENV: PROM_URL")
+	command.Flags().StringVar(&awssecretaccesskey, "awssecretaccesskey", viper.GetString("AWS_SECRET_ACCESS_KEY"), "AWS secret access key for uploading the reports to S3. ENV: AWS_SECRET_ACCESS_KEY")
+	command.Flags().StringVar(&awsaccesskeyid, "awsaccesskeyid", viper.GetString("AWS_ACCESS_KEY_ID"), "AWS key id for uploading the reports to S3. ENV: AWS_ACCESS_KEY_ID")
+	command.Flags().StringVar(&endpointURL, "endpointurl", viper.GetString("ENDPOINT_URL"), "S3 endpoint url for uploading the reports. ENV: ENDPOINT_URL")
+	command.Flags().StringVar(&bucket, "bucket", viper.GetString("BUCKET_NAME"), "S3 bucketname for uploading the reports. ENV: BUCKET_NAME")
+	command.Flags().StringVar(&mimirOrg, "mimirorg", "", "Set the X-Scope-OrgID header for mimir queries")
+	command.Flags().BoolVar(&previousMonth, "previousmonth", false, "Run the report for the previous month. Sets the date to the last day of the previous month and range to 30d. This takes precedence over date and range.")
+
+	return command
+}
+
+func (c *controller) runReport(cmd *cobra.Command, _ []string) error {
+
+	l := log.FromContext(cmd.Context())
+
+	setPreviousMonth()
+
+	metrics, err := slareport.RunQuery(cmd.Context(), promURL, timeRange, date, mimirOrg)
+	if err != nil {
+		return err
+	}
+
+	uploader := slareport.PDFUploader{}
+	err = uploader.Login(cmd.Context(),
+		endpointURL,
+		bucket,
+		awsaccesskeyid,
+		awssecretaccesskey,
+	)
+	if err != nil {
+		return err
+	}
+
+	for customer, metrics := range metrics {
```

Comment:
Nitpick: Shadowing `metrics` is IMO a bit confusing

Reply:
Naming is hard :D